src/main/java/org/springframework/samples/petclinic/rest/importcsv/ImportCSV.java

```diff
@@ -31,109 +30,134 @@
+            String petName = toPetName(field[0]);
+            Date date = toDate(field[1]);
+            PetType petType = toPetType(field[2]);
+            Owner owner = toOwner(field[3]);
+            String operation = toOperation(field[4]);
+            entries.add(new Entry(petName, date, petType, owner, operation));
+        }
+        return entries;
+    }
 
-            field = "";
-            while (i < csv.length() && csv.charAt(i) != ';') {
-                field += csv.charAt(i++);
-            }
-            i++;
-
-            if (pet != null) {
-                ArrayList<PetType> ts = (ArrayList<PetType>) clinicService.findPetTypes();
-                for (int j = 0; j < ts.size(); j++) {
-                    if (ts.get(j).getName().toLowerCase().equals(field)) {
-                        pet.setType(ts.get(j));
-                        break;
-                    }
-                }
-            }
+    public ResponseEntity<List<Pet>> importPets(@RequestBody String csv) {
 
-            field = "";
-            while (i < csv.length() && (csv.charAt(i) != ';' && csv.charAt(i) != '\n')) {
-                field += csv.charAt(i++);
-            }
+        int i = 0;
+        List<Pet> pets = new LinkedList<Pet>();
+        Pet pet;
+        String[] lines = parseCSV(csv);
+        List<Entry> entries = createEntries(lines);
+
+        for (Entry entry : entries) {
+            // we assume pet now accepts an entry, and set its attributes accordingly, e.g. : pet.setName(entry.petName)
+            pet = new Pet();
+            pet.setName(entry.petName);
+            pet.setBirthDate(entry.date);
+            pet.setType(entry.petType);
+            pet.setOwner(entry.ownerName);
 
-            if (pet != null) {
-                String owner = field;
-                List<Owner> matchingOwners = clinicService.findAllOwners()
-                    .stream()
-                    .filter(o -> o.getLastName().equals(owner))
-                    .collect(Collectors.toList());
-
-                if (matchingOwners.size() == 0) {
-                    HttpHeaders headers = new HttpHeaders();
-                    headers.add("errors", "Owner not found");
-                    return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
-                }
-                if (matchingOwners.size() > 1) {
-                    HttpHeaders headers = new HttpHeaders();
-                    headers.add("errors", "Owner not unique");
-                    return new ResponseEntity<List<Pet>>(headers, HttpStatus.BAD_REQUEST);
-                }
-                pet.setOwner(matchingOwners.iterator().next());
+            switch (entry.operation.toLowerCase()) {
+                case "add":
+                    clinicService.savePet(pet);
+                case "delete":
+                    // originally provided logic can be wrapped in this function
+                    clinicService.deletePet(pet);
```

Comment:
`clinicService.deletePet` deletes only pets that are saved in the database/repository, and the argument you pass that you created in line 92 is not in the database. You should instead find the pet object from its owner and delete this object instead.